### PR TITLE
feat(mito): Init the write cache in datanode

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR initializes the `WriteCache` according to the mito config. By default, the write cache is under `DATA_HOME/write_cache` and is disabled.

It addresses an issue that the file cache doesn't use `get()` to access cache entries.

It also refactors some mods to make the integration easier
- Adds a builder for the `CacheManager`
- Recovers the `FileCache` in `WriteCache::new`
- `MitoEngine::new` is an async function now
- Provides a util to attach layers to `ObjectStore`

Now write cache is still an experimental feature so I added prefixes to its config items.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)
- https://github.com/GreptimeTeam/greptimedb/issues/2965

### CODE REVIEW

File: src/datanode/src/datanode.rs [hunk 1]
```diff
@@ -458,20 +458,33 @@ impl DatanodeBuilder {
     async fn build_mito_engine(
         opts: &DatanodeOptions,
         object_store_manager: ObjectStoreManagerRef,
-        config: MitoConfig,
+        mut config: MitoConfig,
     ) -> Result<MitoEngine> {
+        // Sets write cache path if it is empty.
+        if config.experimental_write_cache_path.is_empty() {
+            config.experimental_write_cache_path = join_dir(&opts.storage.data_home, "write_cache");
```
Comment (zhongzc) [inline src/datanode/src/datanode.rs:465]: > Write cache is disabled if the path is empty.

But you modify it...

Comment (evenyag) [reply src/datanode/src/datanode.rs:465]: I struggled with it. I want to put the write cache under our data home by default. We don't know what data_home is so I can't give it a default value in `MitoConfig::default()`. However, others can all `MitoEngine::new()` with an empty write path. At that time I couldn't decide the write cache path and must disable the cache.

Comment (evenyag) [reply src/datanode/src/datanode.rs:465]: Added an enable flag to control the write cache.

File: src/mito2/src/cache.rs [hunk 2]
```diff
@@ -201,6 +147,95 @@ impl CacheManager {
...
+            let cache = Cache::builder()
+                .max_capacity(self.page_cache_size)
+                .weigher(page_cache_weight)
+                .eviction_listener(|k, v, _cause| {
+                    let size = page_cache_weight(&k, &v);
+                    CACHE_BYTES.with_label_values(&[PAGE_TYPE]).sub(size.into());
+                })
+                .build();
+            Some(cache)
+        };
```
Comment (zhongzc) [inline src/mito2/src/cache.rs:228]: style(nonblocking):

```rust
(self.page_cache_size != 0).then(|| Cache::builder()
  .max_capacity(self.page_cache_size)
  .weigher(page_cache_weight)
  .eviction_listener(|k, v, _cause| {
      let size = page_cache_weight(&k, &v);
      CACHE_BYTES.with_label_values(&[PAGE_TYPE]).sub(size.into());
  })
  .build())
```

**Review by zhongzc — APPROVED** [review]

**Review by QuenKar — APPROVED** [review]
LGTM

**Pull request merged**